retry.go

```diff
@@ -44,6 +49,15 @@ type requestGUIDReplacer interface {
 	replace() *url.URL
 }
 
+// Get the request ID from the context if specified, otherwise generate one
+func getRequestID(ctx context.Context) string {
```

Comment:
This going to generate a new request ID every time it's called which isn't idiomatic for context getters. You should be able to call a context getter from anywhere and get the most recent value.

So for example if you did this:

```
func someRestfulFunc(ctx context.Context) {
  requestID := getRequestID(ctx) // returns "foo"
  someOtherRestfulFunc(ctx)
}

func someOtherRestfulFunc(ctx context.Context) {
  requestID := getRequestID(ctx) // uh oh, returns a new request ID
}
```